File: examplepatches/ajstrick81/morphe-androidtv-patches/extensions/extension/src/main/java/ajstrick81/morphe/extension/primevideo/gmb/SendDiagnosticPatch.java
package ajstrick81.morphe.extension.primevideo.gmb;

import android.util.Log;

/**
 * GMB send-direction diagnostic extension — temporary, not for production use.
 *
 * Logs every outbound GMB event type and payload passed into
 * GMBMessageSender.sendGMBMessageToClient() — the sole chokepoint for
 * Java→native GMB traffic — to Android logcat under the tag
 * "GMB_SEND_DIAGNOSTIC" using standard android.util.Log.d(), which surfaces
 * in ADB logcat, unlike Amazon's internal logger.
 *
 * This is a pure diagnostic tool. It does not suppress any messages. Every
 * GMB message passes through normally after being logged.
 *
 * Filter ADB logcat with:
 *   adb logcat --pid=<PID> -v time | findstr "GMB_SEND_DIAGNOSTIC"
 *
 * Mirrors the now-removed receive-direction diagnostic (GMBMessageProcessor
 * .processMessage), which confirmed zero ad-related traffic. This tests the
 * opposite direction. Intended to be removed once the send direction has
 * been confirmed ad-relevant or ad-irrelevant.
 */
@SuppressWarnings("unused")
public class SendDiagnosticPatch {

    private static final String TAG = "GMB_SEND_DIAGNOSTIC";

    /**
     * Called at index 0 of GMBMessageSender.sendGMBMessageToClient().
     * Logs the event type and payload to Android logcat.
     *
     * @param eventType The GMB event type string (e.g. "billing.reportPurchaseEvent")
     * @param payload   The JSON payload for this event
     */
    public static void logGMBSend(String eventType, String payload) {
        try {
            Log.d(TAG, "[TYPE] " + eventType);
            // Log payload separately — payloads can be large JSON strings
            // Truncate to 500 chars to keep logcat readable
            if (payload != null && payload.length() > 0) {
                String truncated = payload.length() > 500
                    ? payload.substring(0, 500) + "...[TRUNCATED]"
                    : payload;
                Log.d(TAG, "[PAYLOAD] " + truncated);
            }
        } catch (Exception e) {
            // Silent fail — never interfere with original sendGMBMessageToClient flow
        }
    }
}

File: examplepatches/ajstrick81/morphe-androidtv-patches/patches/src/main/kotlin/ajstrick81/morphe/patches/primevideo/gmb/SendDiagnosticPatch.kt
package ajstrick81.morphe.patches.primevideo.gmb

import app.morphe.patcher.extensions.InstructionExtensions.addInstructions
import app.morphe.patcher.patch.bytecodePatch
import ajstrick81.morphe.patches.primevideo.misc.extension.primeVideoExtensionPatch
import ajstrick81.morphe.patches.primevideo.shared.Constants

// ─────────────────────────────────────────────────────────────────────────────
// Diagnostic-only — GMB send-direction logging (Java→native)
//
// Not an ad-suppression patch. Logs every eventType/payload passed into
// GMBMessageSender.sendGMBMessageToClient() to standard Android logcat, so
// the send direction of the GMB bus can be empirically checked against ad
// playback — the same question the now-removed processMessage diagnostic
// answered for the receive direction (result: zero ad-related traffic).
//
// Temporary/non-production: intended to be removed once the send direction
// has been confirmed ad-relevant or ad-irrelevant.
// ─────────────────────────────────────────────────────────────────────────────
@Suppress("unused")
val gmbSendDiagnosticPatch = bytecodePatch(
    name = "GMB send diagnostic (debug)",
    description = "Diagnostic-only: logs outbound GMB messages (Java→native) to logcat to test whether the send direction carries ad-related traffic.",
) {
    compatibleWith(Constants.COMPATIBILITY)

    dependsOn(primeVideoExtensionPatch)

    execute {

        // ─────────────────────────────────────────────────────────────────────
        // Hook — GMBMessageSender.sendGMBMessageToClient(String, String, long)
        //
        // Passes eventType/payload straight through to the extension logger
        // and lets the original method body run unmodified afterward — pure
        // observation, no suppression, no register juggling beyond the call.
        // ─────────────────────────────────────────────────────────────────────
        SendGMBMessageToClientFingerprint.method.addInstructions(
            0,
            """
                invoke-static {p1, p2}, Lajstrick81/morphe/extension/primevideo/gmb/SendDiagnosticPatch;->logGMBSend(Ljava/lang/String;Ljava/lang/String;)V
            """
        )
    }
}
